File: src/main/java/kroppeb/server/command/commands/EffectCommand.java
/*
* Copyright (c) 2020 MadHau5
* This Source Code Form is subject to the terms of the Mozilla Public
* License, v. 2.0. If a copy of the MPL was not distributed with this
* file, You can obtain one at http://mozilla.org/MPL/2.0/.
*/

package kroppeb.server.command.commands;

import kroppeb.server.command.Command;
import kroppeb.server.command.arguments.Selector;
import kroppeb.server.command.reader.Reader;
import kroppeb.server.command.reader.ReaderException;
import net.minecraft.entity.Entity;
import net.minecraft.entity.LivingEntity;
import net.minecraft.entity.effect.StatusEffect;
import net.minecraft.entity.effect.StatusEffectInstance;
import net.minecraft.server.command.ServerCommandSource;
import net.minecraft.util.registry.Registry;

import java.util.Collection;

abstract public class EffectCommand implements Command {
	final Selector targets;
	final StatusEffect effect;
	
	public static EffectCommand read(Reader reader) throws ReaderException {
		String sub = reader.readLiteral();
		switch (sub) {
			case "give":
				return GiveCommand.read(reader);
			case "clear":
				return ClearCommand.read(reader);
			default:
				throw new ReaderException("Unexpected effect literal: " + sub);
		}
		
	}
	
	public static class GiveCommand extends EffectCommand {
		final int duration;
		final int amplifier;
		final boolean hideParticles;
		
		public GiveCommand(Selector targets, StatusEffect effect, int duration, int amplifier, boolean hideParticles) {
			this.targets = targets;
			this.effect = effect;
			this.duration = duration;
			this.amplifier = amplifier;
			this.hideParticles = hideParticles;
		}
		
		public static GiveCommand read(Reader reader) throws ReaderException {
			Selector targets = Selector.read(reader);
			StatusEffect effect = Registry.STATUS_EFFECT.get(reader.readIdentifier());
			int duration;
			int amplifier = 0;
			boolean hideParticles = false;
			if (reader.hasNext()) {
				reader.moveNext();
				duration = reader.readInt();
			} else {
				duration = -1;
			}
			if (reader.hasNext()) {
				reader.moveNext();
				amplifier = reader.readInt();
			}
			if (reader.hasNext()) {
				String bool = reader.readLiteral();
				switch (bool) {
					case "true":
						hideParticles = true;
						break;
					case "false":
						hideParticles = false;
						break;
					default:
						throw new ReaderException("Unexpected boolean literal: " + bool);
				}
			}
			return new GiveCommand(targets, effect, duration, amplifier, hideParticles);
		}
		
		public int execute(ServerCommandSource source) {
			int e = 0;
			Collection<Entity> entities = targets.getEntities(source);
			for (Entity entity : entities) {
				if (entity instanceof LivingEntity) {
					LivingEntity livingEntity = (LivingEntity) entity;
					if (effect.isInstant()) {
						if (duration == -1) {
							duration = 1;
						}
					} else {
						if (duration == -1) {
							duration = 30;
						}
						duration *= 20;
					}
					livingEntity.addStatusEffect(new StatusEffectInstance(effect,
							duration, amplifier, false, hideParticles));
					e++;
				}
			}
			if (e == 0) {
				// TODO: throw error on failed command
			}
			return e;
		}
	}
	
	public static class ClearCommand extends EffectCommand {
		
		public ClearCommand(Selector targets, StatusEffect effect) {
			this.targets = targets;
			this.effect = null;
		}
		
		public static ClearCommand read(Reader reader) throws ReaderException {
			Selector targets = Selector.read(reader);
			StatusEffect effect;
			if (reader.hasNext()) {
				effect = Registry.STATUS_EFFECT.get(reader.readIdentifier());
			}
			return new ClearCommand(targets, effect);
		}
		
		public int execute(ServerCommandSource source) {
			int e = 0;
			Collection<Entity> entities = targets.getEntities(source);
			if (effect == null) {
				for (Entity entity : entities) {
					if (entity instanceof LivingEntity) {
						((LivingEntity) entity).clearStatusEffects();
						e++;
					}
				}
			} else {
				for (Entity entity : entities) {
					if (entity instanceof LivingEntity) {
						((LivingEntity) entity).removeStatusEffect(effect);
						e++;
					}
				}
			}
			if (e == 0) {
				// TODO: throw error on failed command
			}
			return e;
		}
	}
}

File: src/main/kotlin/kroppeb/server/command/commands/EffectCommand.kt
/*
 * Copyright (c) 2021 Kroppeb
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/.
 */

package kroppeb.server.command.commands

import kroppeb.server.command.Command
import kroppeb.server.command.InvocationError
import kroppeb.server.command.arguments.selector.PlayerSelector
import kroppeb.server.command.arguments.selector.Selector
import kroppeb.server.command.arguments.selector.SinglePlayerSelector
import kroppeb.server.command.reader.*
import net.minecraft.entity.Entity
import net.minecraft.entity.LivingEntity
import net.minecraft.entity.effect.StatusEffect
import net.minecraft.entity.effect.StatusEffectInstance
import net.minecraft.server.command.ExperienceCommand.Component
import net.minecraft.server.command.ServerCommandSource
import net.minecraft.text.TranslatableText
import net.minecraft.util.registry.Registry


sealed class EffectCommand(val selector: Selector) : Command {

	companion object : ReadFactory<EffectCommand> {

		override fun Reader.parse(): EffectCommand {
			when (val sub = Literal()) {
				"clear" -> {
					if (!canRead())
						return ClearAll(Selector.Self)
					val targets = Selector()
					if (!canRead())
						return ClearAll(targets)
					val id = Id()
					val effect = Registry.STATUS_EFFECT[id] ?: throw ReaderException("Not a valid effect: $id")
					return ClearEffect(targets, effect)
				}


				"give" -> {
					val targets = Selector()
					val id = Id()
					val effect = Registry.STATUS_EFFECT[id] ?: throw ReaderException("Not a valid effect: $id")
					if (!canRead())
						return GiveEffect(targets, effect, null, 0, true)
					val duration = Int()
					if (!canRead())
						return GiveEffect(targets, effect, duration, 0, true)
					val amplifier = Int()
					if (!canRead())
						return GiveEffect(targets, effect, duration, amplifier, true)
					val particles = Boolean()
					return GiveEffect(targets, effect, duration, amplifier, particles)

				}
				else -> expected("effect", "(clear|give)", sub)
			}
		}
	}

	class ClearAll(selector: Selector) : EffectCommand(selector) {
		override fun execute(source: ServerCommandSource): Int {
			val targets = selector.getEntities(source)
			val i = targets.count { it is LivingEntity && it.clearStatusEffects() }

			if (i == 0)
				throw InvocationError()
			return i
		}

	}

	class ClearEffect(selector: Selector, val effect: StatusEffect) : EffectCommand(selector) {
		override fun execute(source: ServerCommandSource): Int {
			val targets = selector.getEntities(source)
			val i = targets.count { it is LivingEntity && it.removeStatusEffect(effect) }
			if (i == 0)
				throw InvocationError()
			return i
		}

	}


	class GiveEffect private constructor(
		selector: Selector,
		val effect: StatusEffect,
		val duration: Int,
		val amplifier: Int,
		val showParticles: Boolean
	) : EffectCommand(selector) {
		constructor(
			selector: Selector,
			effect: StatusEffect,
			duration: Int?,
			amplifier: Int,
			showParticles: Boolean
		) : this(
			selector,
			effect,
			when {
				duration != null -> if (effect.isInstant) duration else duration * 20
				effect.isInstant -> 1
				else -> 600
			},
			amplifier,
			showParticles
		)

		override fun execute(source: ServerCommandSource): Int {
			val targets = selector.getEntities(source)
			val i = targets.count { entity ->
				entity is LivingEntity &&
					entity.addStatusEffect(StatusEffectInstance(effect, duration, amplifier, false, showParticles))
			}

			if (i == 0) throw InvocationError()

			return i
		}

	}

}
